src/database/database-mariadb.cpp

```diff
@@ -0,0 +1,1407 @@
+		throw DatabaseException("MariaDB Error");
+	}
+}
+
+
+void Database_MariaDB::rollback() {
+	try {
+
+		stmtRollback->execute();
+	
+	} catch (sql::SQLException &e) {
+		errorstream << "[MariaDB] Error: " << + e.what() << std::endl;
+		throw DatabaseException("MariaDB Error");
+	}
+}
+
+
+/**
+ * 
+ * 
+ * MapDatabaseMariaDB -- The MariaDB Map Database Backend
+ * 
+ * 
+ */
+
+
+MapDatabaseMariaDB::MapDatabaseMariaDB(const std::string &connect_string) : Database_MariaDB(connect_string, ""), MapDatabase() {
+
+	// connect to database
+	connect();
+
+	// create table if it does not exist
+	createTable(
+		"blocks",
+		"x INT, "
+		"y INT, "
+		"z INT, "
+		"data BLOB, "
+		"PRIMARY KEY (x, y, z)"
+	);
+
+	try {
+
+		// prepare statements
+		stmtDeleteBlock.reset(prepareStatement(
+			"DELETE FROM blocks WHERE x = ? and y = ? and z = ?"
+		));
+
+		stmtLoadBlock.reset(prepareStatement(
+			"SELECT * FROM blocks WHERE x = ? and y = ? and z = ?"
+		));
+
+		stmtListAllLoadableBlocks.reset(prepareStatement(
+			"SELECT x, y, z FROM blocks"
+		));
+
+		stmtSaveBlock.reset(prepareStatement(
+			"INSERT INTO blocks (x, y, z, data) VALUES (?, ?, ?, ?) "
+			"ON DUPLICATE KEY UPDATE data = VALUE(data)"
+		));
+
+	} catch (sql::SQLException &e) {
+		errorstream << "[MariaDB] Error: " << + e.what() << std::endl;
+		throw DatabaseException("MariaDB Error");
+	}
+}
+
+
+bool MapDatabaseMariaDB::deleteBlock(const v3s16 &pos) {
+
+	checkConnection();
+	
+	try {
+
+		stmtDeleteBlock->setInt(1, pos.X);
+		stmtDeleteBlock->setInt(2, pos.Y);
+		stmtDeleteBlock->setInt(3, pos.Z);
+		stmtDeleteBlock->executeUpdate();
+		stmtDeleteBlock->clearParameters();
+
+		std::int32_t rows_affected = stmtDeleteBlock->getUpdateCount();
+
+		return (rows_affected > 0);
+
+	} catch (sql::SQLException &e) {
+		errorstream << "[MariaDB] Error: " << + e.what() << std::endl;
+		throw DatabaseException("MariaDB Error");
+	}
+}
+
+
+void MapDatabaseMariaDB::listAllLoadableBlocks(std::vector<v3s16> &dst) {
+
+	checkConnection();
+
+	try {
+
+		std::unique_ptr<sql::ResultSet> resBlocks(stmtListAllLoadableBlocks->executeQuery());
+
+		while (resBlocks->next()) {
+		
+			v3s16 row(resBlocks->getInt("x"), resBlocks->getInt("y"), resBlocks->getInt("z"));
+			dst.push_back(row);
+		
+		}
+		
+		resBlocks->close();
+
+	} catch (sql::SQLException &e) {
+		errorstream << "[MariaDB] Error: " << + e.what() << std::endl;
+		throw DatabaseException("MariaDB Error");
+	}
+}
+
+
+void MapDatabaseMariaDB::loadBlock(const v3s16 &pos, std::string *block) {
+
+	checkConnection();
+
+	try {
+
+		stmtLoadBlock->setInt(1, htonl(pos.X));
+		stmtLoadBlock->setInt(2, htonl(pos.Y));
+		stmtLoadBlock->setInt(3, htonl(pos.Z));
+		
+		std::unique_ptr<sql::ResultSet> resBlock(stmtLoadBlock->executeQuery());
+
+		if (resBlock->next()) {
+
+			std::stringstream in;
+			in << resBlock->getBlob("data")->rdbuf();
+			std::string data = std::move(in).str();
+			*block = data;
```

Comment:
This makes two copies of the block stream, one in data and one in *block.

Reply:
Fixed in latest commit.